# basic programming
### 1. swapping of two numbers

In [ ]:
a= 10
b= 20
print( a, b)
a,b = b,a    # without using third variable
# temp = a   # using third variable
# a = b
# b = temp
print( a, b)

10 20
20 10


### 2. WAP to find factorial of given number 

algo

- take the input
- iterate using for loop
- till the given input

starting from 1

In [12]:
n = int(input("enter number: "))
# fact =1
# for i in range(1,n+1):
#     fact = fact*i
#     print(f"{fact} x {i}")
# print(fact)

print("using while loop")
i=1
fact=1
while i<=n:
    fact= fact*i
    i+=1
print(fact)

using while loop
120


### 2. WAP to check the given number is a amstrong number or not

- take a number
- extract last digit from given numbers 
- find the cube if the numbers
- add the res to a local variable
- delete the last number from the given number

In [11]:
num = int(input(" enter a number : "))
temp = num
res =0
while num!=0:
    rem = num%10
    res = res+ rem**3
    num = num//10

if res ==temp:
    print("amstrong number")
else:
    print("not amstrong number")

not amstrong number


### 3. WAP TO CHECK the given number is strong or not

- Algorithm for Strong Number
- A strong number is a number whose sum of factorials of digits equals the number itself

Steps:
1. Take a number as input
2. Extract each digit from the number
3. Find the factorial of each digit
4. Add all the factorials together
5. Compare the sum with the original number
6. If sum equals the original number, it's a strong number
7. Otherwise, it's not a strong number

- Example: 145
-  1! + 4! + 5! = 1 + 24 + 120 = 145 (Strong number)
- Example: 134
-  1! + 3! + 4! = 1 + 6 + 24 = 31 ≠ 134 (Not a strong number)

In [ ]:
def factorial(n):
    fact =1
    for i in range(1,n+1):
        fact = fact*i
        # print(f"{fact} x {i}")
    return fact


num = int(input(" enter number "))
temp = num
res=0
while num!=0:
    rem = num%10
    res += factorial(rem)
    num = num//10
if res == temp:
    print("strong number")
else:
    print("not strong number")


### Algorithm for Perfect Number

A perfect number is a positive integer that is equal to the sum of its proper divisors (divisors excluding the number itself).

**Steps:**
1. Take a number as input
2. Find all divisors of the number (excluding the number itself)
3. Calculate the sum of all divisors
4. Compare the sum with the original number
5. If sum equals the original number, it's a perfect number
6. Otherwise, it's not a perfect number

**Examples:**
- 6: Divisors are 1, 2, 3 → 1 + 2 + 3 = 6 (Perfect number)
- 28: Divisors are 1, 2, 4, 7, 14 → 1 + 2 + 4 + 7 + 14 = 28 (Perfect number)
- 10: Divisors are 1, 2, 5 → 1 + 2 + 5 = 8 ≠ 10 (Not a perfect number)

In [15]:
n = int(input(" enter number of terms "))
res = 0 
for i in range(1,n):
    if n%i ==0:
        res += i

if res == n:
    print("perfect number")
else:
    print("not perfect number")

not perfect number


### WAP give all the perfet number in the given range


In [17]:
def perfect_number(n):
    res = 0 
    for i in range(1,n):
        if n%i ==0:
            res += i

    if res == n:
        return True
    else:
        return False
    

num = int(input(" enter number of terms "))
for i in range(1,num+1):
    if perfect_number(i):
        print(f"{i} is perfect number")

6 is perfect number


### Algorithm for Sum of Digits

**Steps:**
1. Take a number as input
2. Extract the last digit using modulo operator (%)
3. Add the digit to a sum variable
4. Remove the last digit by integer division (//)
5. Repeat steps 2-4 until the number becomes 0
6. Print the final sum

**Example:**
- 123: 3 + 2 + 1 = 6
- 456: 4 + 5 + 6 = 15


In [20]:
n = int(input(" enter number"))
sum =0
while n>0:
    digit = n % 10
    sum = sum + digit
    n = n // 10
print("Sum of digits:", sum)


Sum of digits: 6


### WAP to reverse the given number

**Steps:**
1. Take a number as input
2. Extract the last digit using modulo operator (%)
3. Multiply the reversed number by 10 and add the extracted digit
4. Remove the last digit by integer division (//)
5. Repeat steps 2-4 until the number becomes 0
6. Print the reversed number

**Example:**
- 123: reverse is 321
- 4567: reverse is 7654

In [1]:
n = int(input(" enter number"))
rev =0
while n>0:
    digit  = n % 10
    rev = rev * 10 + digit
    n = n // 10
print("Reversed number:", rev)

Reversed number: 321


### wap to CHECK the given number is special number or not 

special number is the each digit sum and each digit product is added then we will get the given number

**Steps:**
1. Take a number as input
2. Extract each digit from the number
3. Calculate the sum of all digits
4. Calculate the product of all digits
5. Add the sum and product together
6. Compare the result with the original number
7. If the result equals the original number, it's a special number
8. Otherwise, it's not a special number

**Example:**
- 10: Digits are 1, 0 → Sum = 1 + 0 = 1, Product = 1 × 0 = 0 → 1 + 0 = 1 ≠ 10 (Not a special number)
- 132: Digits are 1, 3, 2 → Sum = 1 + 3 + 2 = 6, Product = 1 × 3 × 2 = 6 → 6 + 6 = 12 ≠ 132 (Not a special number)
- 19: Digits are 1, 9 → Sum = 1 + 9 = 10, Product = 1 × 9 = 9 → 10 + 9 = 19 (Special number)

In [3]:
n = int(input(" enter number"))
s, p =0,1
temp = n
while n>0:
    digit = n % 10
    s += digit
    p *= digit
    n = n // 10

if temp == s + p:
    print("spy number")
else:
    print("not spy number") 

not spy number


### wa program to check given number is a dusarium number or not

A Disarium number is a number where the sum of its digits powered with their respective positions is equal to the original number.

**Algorithm for Disarium Number:**

1. Take a number as input
2. Extract each digit from the number one by one
3. Raise each digit to the power of its position (1-indexed)
4. Add all the powered digits together
5. Compare the sum with the original number
6. If sum equals the original number, it's a Disarium number
7. Otherwise, it's not a Disarium number

**Examples:**
- 135: 1¹ + 3² + 5³ = 1 + 9 + 125 = 135 (Disarium number)
- 175: 1¹ + 7² + 5³ = 1 + 49 + 125 = 175 (Disarium number)
- 124: 1¹ + 2² + 4³ = 1 + 4 + 64 = 69 ≠ 124 (Not a Disarium number)

In [5]:
n = int(input(" enter number"))
l = len(str(n))
temp = n
sum = 0
while n>0:
    digit  = n % 10
    sum += digit ** l
    l -= 1
    n = n // 10

if temp == sum:
    print("disarium number")
else:
    print("not disarium number")

not disarium number


### wap to find the frequency of the number

**Steps:**
1. Take a number as input
2. Initialize a frequency dictionary
3. Extract each digit from the number
4. Update the frequency count for each digit in the dictionary
5. Print the frequency of each digit


In [6]:
n = int(input(" enter number"))

d = {}
while n>0:
    digit = n%10
    if digit not in d:
        d[digit] = 1
    else:
        d[digit] += 1
    n = n//10

for key, value in d.items():
    print(f"Digit {key} has frequency {value}")

Digit 2 has frequency 1
Digit 7 has frequency 1
Digit 6 has frequency 2
Digit 5 has frequency 3
Digit 3 has frequency 3
Digit 4 has frequency 2


### wa program to check automorphic number or not

**Steps:**
1. Take a number as input
2. Calculate the square of the number   
3. Check if the square ends with the same digits as the original number
4. If it does, it's an automorphic number
5. Otherwise, it's not an automorphic number

**Examples:**
- 5: 5² = 25 (Automorphic number)
- 6: 6² = 36 (Automorphic number)
- 7: 7² = 49 (Not an automorphic number)
- 25: 25² = 625 (Automorphic number)

In [ ]:
n = int(input(" enter number"))
square = n * n
temp = n
l = 0
while temp > 0:
    l += 1
    temp //= 10

divisor = 1
for i in range(l):
    divisor *= 10

last_digits = square % divisor
if last_digits == n:
    print("automorphic number")
else:
    print("not automorphic number")


not automorphic number


### wa program to check the given number is palindrome or not

**Steps:**
1. Take a number as input
2. Extract each digit from the number
3. Build the reversed number
4. Compare the reversed number with the original number
5. If they are the same, it's a palindrome
**Example:**
- 121: Palindrome
- 456: Not a palindrome

In [ ]:
n = int(input(" enter number"))
temp = n
rev = 0 
while n>0:
    digit = n % 10
    rev = rev * 10 + digit
    n = n // 10

if rev == temp:
    print("palindrome number")
else:
    print("not palindrome number")

1
2
1
palindrome number


### wa program to check the given number is evil or odious number

**Steps:**
1. Take a number as input
2. Convert the number to its binary representation
3. Count the number of 1's in the binary representation
4. If the count of 1's is even, it's an evil number
5. If the count of 1's is odd, it's an odious number

- Example:
- 3 (binary 11) has two 1's → Evil number
- 5 (binary 101) has two 1's → Evil number
- 7 (binary 111) has three 1's → Odious number
- 10 (binary 1010) has two 1's → Evil number 

In [ ]:
n = int(input(" enter number"))
rem =0
count =0
temp = 1
res =0
while n>0:
    rem = n % 2
    # print(rem)
    res = res + temp * rem
    # print(res)
    if rem == 1:
        count += 1
    n = n // 2
    temp = temp * 10
print("Binary equivalent:", res )
if count%2 ==0:
    print("evil number")
else:
    print("odious number")
    

0
10
10
1010
Binary equivalent: 1010
evil number


### WAP to shift all odd number to the first place and even numbers to the last place in a given number

**Steps:**
1. Take a number as input
2. Extract each digit from the number
3. Separate odd and even digits
4. Combine odd digits followed by even digits
5. Print the rearranged number
- Example:
- Input: 123456
- Output: 135246




In [ ]:

n = int(input(" enter number: "))
if n == 0:
    print("Number after rearranging:", 0)
else:
    temp = n
    odds_rev = evens_rev = 0
    odds_count = evens_count = 0

    while temp > 0:
        d = temp % 10
        if d % 2 != 0:
            odds_rev = odds_rev * 10 + d
            odds_count += 1
        else:
            evens_rev = evens_rev * 10 + d
            evens_count += 1
        temp //= 10

    def reverse_num(x):
        r = 0
        while x > 0:
            r = r * 10 + (x % 10)
            x //= 10
        return r

    odds = reverse_num(odds_rev)
    evens = reverse_num(evens_rev)


    result = odds * 10 ** evens_count + evens
    print("Number after rearranging:", result)



Number after rearranging: 135246


### program to check the given number is zuker man number or not
- it is a number which is completely divisible by its product of a given number
**steps:**
1. Take a number as input
2. Extract each digit from the number
3. Calculate the product of all digits
4. Check if the original number is divisible by the product

- input 24
- 2 * 4 = 8
- 24 % 8 == 0



In [3]:
n = int(input(" enter number"))
product = 1
temp = n
while n > 0:
    digit = n % 10
    product = product * digit
    n = n//10
if temp % product == 0:
    print("zukar man number")
else:
    print("not zukar man number")

not zukar man number


### wa program to check given number is a smith number

**Steps:**
1. Take a number as input
2. Find the prime factors of the number
3. Calculate the sum of the digits of the prime factors
4. Calculate the sum of the digits of the original number
5. Compare the two sums
- If both sums are equal, it's a Smith number

- Example:
- 4: Prime factors are 2, 2 → Sum of digits = 2 + 2 = 4 (Smith number)
- smith number examples
- 6: Prime factors are 2, 3 → Sum of digits = 2
- 22: Prime factors are 2, 11 → Sum of digits = 2 + (1 + 1) = 4 ≠ 2 + 2 = 4 (Not a Smith number)
- 27: Prime factors are 3, 3, 3 → Sum of digits = 3 + 3 + 3 = 9 ≠ 2 + 7 = 9 (Not a Smith number)
- 58: Prime factors are 2, 29 → Sum of digits = 2 + (2 + 9) = 13 ≠ 5 + 8 = 13 (Not a Smith number)



In [26]:


def prime_factor_sum(n):
    factor = 2
    sum = 0
    print("Prime factors are: ", end="")
    while n != 1:
        if n % factor == 0:
            sum = sum + factor
            n = n // factor
            print(factor, end=" ")
        else:
            factor +=1
        print(sum)
    return sum
sum2 = 0 
while n>0:
    digit = n%10
    sum2 += digit
    n //= 10
print(sum2)

n = int(input(" enter number"))

if sum2 == prime_factor_sum(n):
    print("smith number")
else:
    print("not smith number")



4
Prime factors are: 2 2
2 4
smith number


### wAp to check the given number is ugly number or not

- it is a number which includes the prime factors like 2 or 3 or 5 only,  if it is getting any other prime factor then it is not an ugly number
- number which contains only the prime factors=> 2 3 5

**Steps:**
1. Take a number as input
2. Divide the number by 2, 3, and 5 repeatedly until it is no longer divisible by these numbers
3. If the resulting number is 1, it's an ugly number
4. Otherwise, it's not an ugly number

example:
- 6: Divisible by 2 and 3 → Ugly number
- 14: Divisible by 2 but not by 3 or 5 → Not an ugly number
- 30: Divisible by 2, 3, and 5 → Ugly number
- 25: Divisible by 5 only → Ugly number
- 8: Divisible by 2 only → Ugly number
- 7: Not divisible by 2, 3, or 5 → Not an ugly number



In [30]:
n = int(input(" enter number"))
original_n = n
while n % 2 == 0:
    n = n // 2
print(n)
while n % 3 == 0:
    n = n // 3
print(n)
while n % 5 == 0:
    n = n // 5
print(n)
if n == 1:
    print(f"{original_n} is an ugly number")
else:
    print(f"{original_n} is not an ugly number")

3
1
1
6 is an ugly number


trimorphic number
- it is a number whose cube ends with the same digits as the original number

In [3]:
def reverse_num(x):
    r = 0
    while x > 0:
        r = r * 10 + (x % 10)
        x //= 10
    return r

n = int(input(" enter number"))

p = True

num_rev = 0
qube = n **3

while qube>0:
    digit = qube % 10
    num_rev = num_rev * 10 + digit
    qube = qube //10

    if n == reverse_num(num_rev):
        p = True
        break
    else:
        p = False

if p:
    print("trimorphic number")
else:
    print("not trimorphic number")

not trimorphic number


maximum subarray sum

In [6]:
def max_subarray_sum(arr):
    max_sum = 0
    if len(arr) == 1:
        return arr[0]
    else:
        size = int(input(" enter size of array "))
        for i in range(len(arr)):
            curr_sum = 0
            if i + size >len(arr):
                return max_sum
            else:
                for j in range(i, i + size):
                    curr_sum += arr[j]
            if curr_sum > max_sum:
                max_sum = curr_sum
        return max_sum


l1 = [-2]
print("Maximum subarray sum is:", max_subarray_sum(l1))

Maximum subarray sum is: -2


In [ ]:
l1 = ["flower","flow","flight"]

word = l1[0]
sub_s = ''
for char in word:
    sub_s += char
    if len(l1) == 1:
        print(sub_s)
        break
    else:
        for i in range(1, len(l1)):
            if l1[i].startswith(sub_s):
                continue
            else:
                sub_s = sub_s[:-1]
                break
                
if sub_s != '':
    print("Longest common prefix is:", sub_s)
else:
    print("No common prefix found")

No common prefix found
